In [1]:
# ==============================================================================
# ENVIRONMENT SETUP AND GITHUB AUTHENTICATION WITHOUT GOOGLE DRIVE
# Objective: Configure GitHub authentication and clone or update the repository
# directly inside the temporary Google Colab environment.
# ==============================================================================

# Import os for file-system operations and working-directory management.
import os

# Import userdata to retrieve the GitHub token from Colab Secrets.
from google.colab import userdata


# ------------------------------------------------------------------------------
# 1. PARAMETER CONFIGURATION
# ------------------------------------------------------------------------------

# Define the GitHub account and repository names.
GITHUB_USER = "pascual-francisco"
GITHUB_REPO = "INF-8239-Ciencia-de-Datos-II"

# Retrieve the GitHub personal access token from Colab Secrets.
GITHUB_TOKEN = userdata.get("TOKEN")

# Define the repository location directly inside the Colab environment.
REPO_DIR = f"/content/{GITHUB_REPO}"

# Define the author information used in Git commits.
GIT_AUTHOR_NAME = "Francisco Pascual"
GIT_AUTHOR_EMAIL = "pascual.francisco@gmail.com"


# ------------------------------------------------------------------------------
# 2. VERIFY THE GITHUB TOKEN
# ------------------------------------------------------------------------------

# Stop execution if the TOKEN secret is unavailable.
if not GITHUB_TOKEN:
    raise ValueError(
        "The GitHub token was not found. "
        "Add a Colab Secret named TOKEN before running this cell."
    )


# ------------------------------------------------------------------------------
# 3. GLOBAL GIT CONFIGURATION
# ------------------------------------------------------------------------------

# Configure the author name used in Git commits.
!git config --global user.name "{GIT_AUTHOR_NAME}"

# Configure the author email used in Git commits.
!git config --global user.email "{GIT_AUTHOR_EMAIL}"

# Configure main as the default branch name.
!git config --global init.defaultBranch main


# ------------------------------------------------------------------------------
# 4. DEFINE THE GITHUB REPOSITORY URL
# ------------------------------------------------------------------------------

# Define the authenticated URL used to clone or update the repository.
AUTH_URL = (
    f"https://{GITHUB_USER}:{GITHUB_TOKEN}"
    f"@github.com/{GITHUB_USER}/{GITHUB_REPO}.git"
)

# Define a clean URL that does not contain the GitHub token.
CLEAN_URL = (
    f"https://github.com/{GITHUB_USER}/{GITHUB_REPO}.git"
)


# ------------------------------------------------------------------------------
# 5. CLONE OR UPDATE THE REPOSITORY
# ------------------------------------------------------------------------------

# Clone the repository when it is not available in the current Colab session.
if not os.path.exists(REPO_DIR):
    print(f"Cloning {GITHUB_REPO}...")

    # Clone the repository using the authenticated URL.
    !git clone "{AUTH_URL}" "{REPO_DIR}"

else:
    print("Repository already exists. Updating from the main branch...")

    # Navigate to the existing repository.
    %cd {REPO_DIR}

    # Retrieve and merge the latest changes using authenticated access.
    !git pull "{AUTH_URL}" main


# ------------------------------------------------------------------------------
# 6. REMOVE THE TOKEN FROM THE STORED REMOTE URL
# ------------------------------------------------------------------------------

# Navigate to the repository after cloning or updating it.
%cd {REPO_DIR}

# Store a clean GitHub URL as the permanent origin remote.
# This prevents the token from remaining visible in the Git configuration.
!git remote set-url origin "{CLEAN_URL}"


# ------------------------------------------------------------------------------
# 7. VERIFY THE ACTIVE ENVIRONMENT
# ------------------------------------------------------------------------------

# Display the active repository location.
print(f"\nEnvironment ready and located in:\n{os.getcwd()}")

# Display the configured Git remote without exposing the token.
print("\nConfigured Git remote:")
!git remote -v

# Display the current repository status.
print("\nCurrent Git status:")
!git status --short

Cloning INF-8239-Ciencia-de-Datos-II...
Cloning into '/content/INF-8239-Ciencia-de-Datos-II'...
remote: Enumerating objects: 751, done.
remote: Counting objects: 100% (181/181), done.
remote: Compressing objects: 100% (132/132), done.
remote: Total 751 (delta 91), reused 125 (delta 48), pack-reused 570 (from 1)
Receiving objects: 100% (751/751), 25.49 MiB | 5.47 MiB/s, done.
Resolving deltas: 100% (427/427), done.
/content/INF-8239-Ciencia-de-Datos-II

Environment ready and located in:
/content/INF-8239-Ciencia-de-Datos-II

Configured Git remote:
origin	https://github.com/pascual-francisco/INF-8239-Ciencia-de-Datos-II.git (fetch)
origin	https://github.com/pascual-francisco/INF-8239-Ciencia-de-Datos-II.git (push)

Current Git status:


# U02.LAB04 — Dataset Candidate Comparison

## Proposed Research Question

**Can a Natural Language Processing model classify Spanish-language tweets into negative, neutral, and positive sentiment categories?**

## Candidate Comparison

| Evaluation criterion | Spanish Tweet Datasets Based on Context of Words | TwitterSentimentDataset |
|---|---|---|
| **Source** | Universidad de Alcalá institutional repository | Public GitHub repository |
| **Text language** | Spanish | Spanish |
| **Unit of analysis** | Tweet | Tweet |
| **Text origin** | Naturally occurring Spanish-language tweets | Naturally occurring Spanish-language tweets |
| **Sentiment labels** | Positive (`P`), neutral (`NEU`), negative (`N`), and no opinion (`NONE`) | Positive, negative, and tweets without classified sentiment |
| **Explicit neutral class** | Yes | No |
| **Additional category** | `NONE`, representing no expressed opinion | Tweets without classified sentiment |
| **Approximate size** | 3,413 InterTASS tweets and 57,832 General Corpus tweets | Approximately 250,000 tweets |
| **Annotation method** | Sentiment annotations inherited from TASS and InterTASS | Automatically inferred primarily from positive and negative emoticons |
| **License** | Creative Commons Attribution 4.0 International (`CC BY 4.0`) | Apache License 2.0 |
| **Version identification** | Version 1.0 | Repository version must be identified through a commit |
| **Persistent identifier** | DOI: `10.21950/XEV9IC` | No institutional DOI identified |
| **Reproducibility** | High because the dataset has a DOI, version, institutional source, and documented partitions | Moderate because the repository is public, but the exact version should be pinned to a commit |
| **Class balance** | Severely imbalanced because neutral represents approximately 2.25% of the General Corpus | Must be audited directly |
| **Label quality** | More appropriate for supervised sentiment classification | Weak labeling based on emoticons may introduce noise |
| **Main quality concern** | Severe underrepresentation of neutral sentiment | Absence of an emoticon does not prove neutral sentiment |
| **Main representativeness concern** | Historical Twitter language from specific collection periods | Automatically selected Twitter data may overrepresent users who use emoticons |
| **Risk** | May contain sensitive language, mentions, political opinions, irony, or identifiable references | May contain user-generated content, sensitive language, and incorrectly inferred labels |
| **Required transformation** | Keep `P`, `NEU`, and `N`; exclude `NONE` | A reliable neutral class cannot be obtained directly without additional annotation |
| **Final decision** | **Accepted with conditions** | **Rejected** |

---

## Accepted Dataset

### Spanish Tweet Datasets Encoded for Sentiment Analysis Based on Context of Words

The dataset published by the Universidad de Alcalá was accepted because it contains Spanish-language tweets with explicit positive, neutral, and negative sentiment labels.

The corpus has documented academic provenance, an identifiable version, a persistent DOI, reproducible public access, and a Creative Commons Attribution 4.0 International (`CC BY 4.0`) license.

The dataset includes the following original labels:

- `P`: Positive sentiment
- `NEU`: Neutral sentiment
- `N`: Negative sentiment
- `NONE`: No expressed opinion

For the proposed three-class sentiment-analysis task, the processed dataset will retain `P`, `NEU`, and `N`. The `NONE` category will be excluded because absence of opinion is conceptually different from neutral sentiment.

The principal limitation is the severe underrepresentation of the neutral class. In the General Corpus, neutral observations represent approximately 2.25% of the data. Therefore, accuracy alone may provide a misleading evaluation. Macro F1-score, recall by class, and the confusion matrix should be prioritized.

**Decision: Accepted with conditions.**

---

## Rejected Dataset

### TwitterSentimentDataset

TwitterSentimentDataset was considered because it contains a large number of naturally occurring Spanish-language tweets and is publicly available under the Apache License 2.0.

The corpus contains approximately 250,000 tweets divided into positive, negative, and unclassified-sentiment groups. Positive and negative labels were generated automatically using emoticons as sentiment indicators.

The dataset was rejected for the following reasons:

1. It does not provide an explicit neutral sentiment class.
2. Tweets without a positive or negative emoticon are not necessarily neutral.
3. A tweet can express sentiment without using an emoticon.
4. Emoticons can be used ironically or sarcastically.
5. The automatic labeling procedure may introduce substantial label noise.
6. The corpus may overrepresent users and messages that use emoticons.
7. An institutional DOI and formally documented dataset version were not identified.
8. The proposed research question requires an interpretable neutral class.

Although the dataset is larger, dataset size does not compensate for weak target-label validity. The accepted corpus provides labels that better correspond to the intended three-class classification task.

**Decision: Rejected because it does not provide a reliable neutral class.**

---

# Final Selection Decision

| Decision component | Result |
|---|---|
| **Accepted dataset** | Spanish Tweet Datasets Encoded for Sentiment Analysis Based on Context of Words |
| **Rejected dataset** | TwitterSentimentDataset |
| **Main reason for acceptance** | Explicit positive, neutral, and negative labels with institutional provenance |
| **Main reason for rejection** | The unclassified group cannot be interpreted reliably as neutral sentiment |
| **Accepted dataset license** | CC BY 4.0 |
| **Rejected dataset license** | Apache License 2.0 |
| **Main accepted-dataset limitation** | Severe underrepresentation of neutral sentiment |
| **Required processing decision** | Exclude `NONE` and preserve only `P`, `NEU`, and `N` |

## Sources

- Selected dataset: https://doi.org/10.21950/XEV9IC
- Rejected dataset: https://github.com/garnachod/TwitterSentimentDataset


In [ ]:
# ==============================================================================
# CELL 04 — COMPLETE THE DATASET CANDIDATE COMPARISON
# Objective: Document the accepted and rejected public corpus candidates
# using the instructor-provided CSV structure.
# ==============================================================================

# Import pandas for creating, saving, and displaying the comparison table.
import pandas as pd

# Define the exact path to the dataset candidate comparison file.
candidates_file = (
    "/content/INF-8239-Ciencia-de-Datos-II/"
    "Unit_02_Natural_Language_Processing/"
    "nlp_sentiment_analysis/docs/dataset_candidates.csv"
)

# Create the comparison table using the original column names.
candidates_df = pd.DataFrame([
    {
        "name": "Spanish Tweet Datasets Based on Context of Words",
        "url": "https://doi.org/10.21950/XEV9IC",
        "license": "CC BY 4.0",
        "language": "Spanish",
        "rows": 57832,
        "target": "P, NEU, N, NONE",
        "decision": "accepted"
    },
    {
        "name": "TwitterSentimentDataset",
        "url": "https://github.com/garnachod/TwitterSentimentDataset",
        "license": "Apache 2.0",
        "language": "Spanish",
        "rows": 250000,
        "target": "positive, negative, unclassified",
        "decision": "rejected"
    }
])

# Save the completed table to the instructor-provided CSV file.
# index=False prevents the creation of an extra numerical index column.
candidates_df.to_csv(
    candidates_file,
    index=False,
    encoding="utf-8"
)

# Display the completed dataset candidate comparison.
display(candidates_df)

# Confirm the exact location where the file was saved.
print(f"Dataset candidate comparison saved to:\n{candidates_file}")

,name,url,license,language,rows,target,decision
0,Spanish Tweet Datasets Based on Context of Words,https://doi.org/10.21950/XEV9IC,CC BY 4.0,Spanish,57832,"P, NEU, N, NONE",accepted
1,TwitterSentimentDataset,https://github.com/garnachod/TwitterSentimentD...,Apache 2.0,Spanish,250000,"positive, negative, unclassified",rejected


Dataset candidate comparison saved to:
/content/INF-8239-Ciencia-de-Datos-II/Unit_02_Natural_Language_Processing/nlp_sentiment_analysis/docs/dataset_candidates.csv


In [ ]:
# ==============================================================================
# CELL 05 — VERIFY THE SAVED CANDIDATE COMPARISON
# Objective: Reload the saved CSV file and confirm that its content and
# column structure were preserved correctly.
# ==============================================================================

# Reload the CSV file directly from the project directory.
saved_candidates_df = pd.read_csv(candidates_file)

# Display the saved comparison table.
display(saved_candidates_df)

# Display the preserved column names.
print("Saved columns:")
print(saved_candidates_df.columns.tolist())

,name,url,license,language,rows,target,decision
0,Spanish Tweet Datasets Based on Context of Words,https://doi.org/10.21950/XEV9IC,CC BY 4.0,Spanish,57832,"P, NEU, N, NONE",accepted
1,TwitterSentimentDataset,https://github.com/garnachod/TwitterSentimentD...,Apache 2.0,Spanish,250000,"positive, negative, unclassified",rejected


Saved columns:
['name', 'url', 'license', 'language', 'rows', 'target', 'decision']


# Dataset Candidate Evaluation

## Proposed Research Question

**Can a Natural Language Processing model classify Spanish-language tweets into negative, neutral, and positive sentiment categories?**

Two public Spanish-language corpora were evaluated according to the approval criteria established for U02.LAB04.

---

## Candidate Comparison

| Criterion | Spanish Tweet Datasets Based on Context of Words | TwitterSentimentDataset |
|---|---|---|
| **Relevance** | The corpus contains Spanish-language tweets with explicit positive (`P`), neutral (`NEU`), negative (`N`), and no-opinion (`NONE`) labels. The `P`, `NEU`, and `N` categories directly support the proposed three-class sentiment-classification task. | The corpus contains Spanish-language tweets classified as positive, negative, or unclassified. However, the unclassified group is not an explicitly annotated neutral class. Therefore, the labels do not fully support the proposed three-class task. |
| **License** | The dataset is distributed under the Creative Commons Attribution 4.0 International License (`CC BY 4.0`). The license permits sharing and adaptation, provided that appropriate attribution is given to the original authors and source. | The repository declares the Apache License 2.0. However, the license information should be interpreted carefully because the repository contains user-generated Twitter content in addition to software and documentation. |
| **Representativeness** | The corpus contains naturally occurring Spanish-language tweets. The General Corpus covers topics such as politics, economics, communication, and culture and includes tweets primarily from 2011–2012. InterTASS contains tweets from 2016–2017. The dataset may not represent current Spanish usage, all Spanish-speaking countries, all dialects, or communication outside Twitter. | The corpus contains tweets recognized as Spanish by the Twitter API. Its automatic collection method may overrepresent users who use emoticons. It may not represent formal Spanish, current language, all geographic varieties of Spanish, or users who express opinions without emoticons. |
| **Quality** | The corpus has enough observations for auditing and future model development. However, the neutral class is severely underrepresented. In the General Corpus, `NEU` represents approximately 2.25% of the observations. Missing values, empty texts, duplicates, invalid labels, and cross-partition duplicates must still be measured by the local audit script. | Positive and negative labels were generated automatically using emoticons. This weak-labeling method may introduce classification errors because emoticons can be absent, ambiguous, or ironic. The unclassified group cannot be assumed to be neutral. Missing values, duplicates, class distribution, and text lengths must also be verified locally. |
| **Reproducibility** | The dataset has an institutional source, documented authors, version 1.0, a persistent DOI, an identifiable license, and documented corpus partitions. Another person can locate the same dataset through its DOI, although the exact downloaded file must also be identified using a SHA-256 hash. | The repository is publicly accessible, but no institutional DOI or formal dataset version was identified. Reproducibility would require recording the exact repository URL, commit identifier, download date, file name, and SHA-256 hash. |
| **Risk** | The corpus originates from social media and may contain usernames, mentions, political opinions, offensive language, sensitive statements, or references to identifiable individuals. Examples included in documentation must be anonymized, and raw data should not be unnecessarily republished. | The corpus also contains user-generated social-media content and may include personal references, offensive language, sensitive statements, or other potentially harmful content. Automatic collection and labeling

In [ ]:
# ==============================================================================
# CELL 05 — CONFIGURE THE SAMPLE DATASET SOURCE
# Objective: Configure the project to use the available sample dataset
# with its actual text and target column names.
# ==============================================================================

# Import Path for operating-system-independent path management.
from pathlib import Path

# Define the exact path to the hidden .env configuration file.
env_path = Path(
    "/content/INF-8239-Ciencia-de-Datos-II/"
    "Unit_02_Natural_Language_Processing/"
    "nlp_sentiment_analysis/.env"
)

# Define the configuration using the available sample dataset.
env_content = """DATA_SOURCE=local
DATASET_PATH=data/sample/demo_text.csv
DATASET_URL=
TEXT_COLUMN=text
TARGET_COLUMN=label
"""

# Write the configuration to the hidden .env file.
env_path.write_text(env_content, encoding="utf-8")

# Confirm that the configuration was saved successfully.
print(f"Environment configuration saved to:\n{env_path}")

Environment configuration saved to:
/content/INF-8239-Ciencia-de-Datos-II/Unit_02_Natural_Language_Processing/nlp_sentiment_analysis/.env


In [ ]:
# ==============================================================================
# CELL 06 — VERIFY THE ENVIRONMENT CONFIGURATION
# Objective: Confirm that the .env file contains the correct dataset settings.
# ==============================================================================

# Display the current .env configuration.
print(env_path.read_text(encoding="utf-8"))

DATA_SOURCE=local
DATASET_PATH=data/sample/demo_text.csv
DATASET_URL=
TEXT_COLUMN=text
TARGET_COLUMN=label



In [ ]:
# ==============================================================================
# CELL 07 — DOWNLOAD THE SELECTED PUBLIC CORPUS
# Objective: Download the approved Spanish sentiment corpus from its official
# institutional repository and preserve the original ZIP file in data/raw.
# ==============================================================================

# Import Path for operating-system-independent path management.
from pathlib import Path

# Import urlretrieve to download the public dataset file.
from urllib.request import urlretrieve

# Define the NLP project root directory.
PROJECT_PATH = Path(
    "/content/INF-8239-Ciencia-de-Datos-II/"
    "Unit_02_Natural_Language_Processing/"
    "nlp_sentiment_analysis"
)

# Define the raw data directory.
raw_data_path = PROJECT_PATH / "data" / "raw"

# Define the official Dataverse download URL for the selected dataset file.
dataset_url = (
    "https://edatos.consorciomadrono.es/"
    "api/access/datafile/53509"
)

# Define the destination path without changing the original file name.
dataset_zip_path = raw_data_path / "Data.zip"

# Download the original dataset archive only if it is not already available.
if not dataset_zip_path.exists():
    urlretrieve(dataset_url, dataset_zip_path)
    print(f"Dataset downloaded to:\n{dataset_zip_path}")
else:
    print(f"Dataset archive already exists:\n{dataset_zip_path}")

Dataset downloaded to:
/content/INF-8239-Ciencia-de-Datos-II/Unit_02_Natural_Language_Processing/nlp_sentiment_analysis/data/raw/Data.zip


In [ ]:
# ==============================================================================
# CELL 08 — INSPECT THE DOWNLOADED DATASET ARCHIVE
# Objective: Display the files contained in the original dataset archive
# without extracting or modifying them.
# ==============================================================================

# Import Path for operating-system-independent path management.
from pathlib import Path

# Import ZipFile to inspect the contents of the downloaded ZIP archive.
from zipfile import ZipFile

# Define the NLP project root directory.
PROJECT_PATH = Path(
    "/content/INF-8239-Ciencia-de-Datos-II/"
    "Unit_02_Natural_Language_Processing/"
    "nlp_sentiment_analysis"
)

# Define the path to the original downloaded archive.
dataset_zip_path = PROJECT_PATH / "data" / "raw" / "Data.zip"

# Verify that the ZIP archive exists before inspecting it.
if dataset_zip_path.exists():

    # Open the archive in read-only mode.
    with ZipFile(dataset_zip_path, "r") as dataset_archive:

        # Retrieve the information for every entry stored in the archive.
        archive_files = dataset_archive.infolist()

        # Display the archive name and total number of entries.
        print(f"Archive: {dataset_zip_path.name}")
        print(f"Total entries: {len(archive_files)}")
        print("\nArchive contents:\n")

        # Display the name and uncompressed size of each entry.
        for archive_file in archive_files:
            print(
                f"{archive_file.filename} "
                f"({archive_file.file_size:,} bytes)"
            )

else:
    print(f"Dataset archive not found: {dataset_zip_path}")

Archive: Data.zip
Total entries: 10

Archive contents:

Corpus_gral/ (0 bytes)
Corpus_gral/readme.txt (1,667 bytes)
Corpus_gral/reviews_cod.txt (70,105,325 bytes)
Corpus_gral/reviews_tw_definitivo.txt (621,270 bytes)
Corpus_gral/words_cod.txt (1,012,295 bytes)
Datos_all/ (0 bytes)
Datos_all/readme.txt (1,671 bytes)
Datos_all/reviews_cod_all.txt (2,597,659 bytes)
Datos_all/reviews_tw_all.txt (33,000 bytes)
Datos_all/words_cod_all.txt (141,399 bytes)


In [ ]:
# ==============================================================================
# CELL 09 — VERIFY THE DOWNLOADED DATASET FILE
# Objective: Calculate the SHA-256 hash of the original dataset archive
# to identify the exact downloaded file.
# ==============================================================================

# Import Path for operating-system-independent path management.
from pathlib import Path

# Import hashlib to calculate the SHA-256 hash.
import hashlib

# Define the path to the downloaded dataset archive.
dataset_zip_path = Path(
    "/content/INF-8239-Ciencia-de-Datos-II/"
    "Unit_02_Natural_Language_Processing/"
    "nlp_sentiment_analysis/data/raw/Data.zip"
)

# Create a SHA-256 hashing object.
sha256_hash = hashlib.sha256()

# Read the file in binary chunks and update the hash.
with dataset_zip_path.open("rb") as dataset_file:
    for chunk in iter(lambda: dataset_file.read(8192), b""):
        sha256_hash.update(chunk)

# Display the dataset file information.
print(f"File name: {dataset_zip_path.name}")
print(f"File size: {dataset_zip_path.stat().st_size:,} bytes")
print(f"SHA-256: {sha256_hash.hexdigest()}")

File name: Data.zip
File size: 10,699,414 bytes
SHA-256: 42709113f5b10ef421ad0b900f6b836f28fa02073ca3d0869edb689bc9c825d2


In [ ]:
# ==============================================================================
# CELL 10 — EXTRACT THE DOWNLOADED DATASET ARCHIVE
# Objective: Extract the dataset files into a separate raw-data directory
# while preserving the original ZIP archive without modification.
# ==============================================================================

# Import Path for operating-system-independent path management.
from pathlib import Path

# Import ZipFile to extract the downloaded dataset archive.
from zipfile import ZipFile

# Define the original downloaded archive.
dataset_zip_path = Path(
    "/content/INF-8239-Ciencia-de-Datos-II/"
    "Unit_02_Natural_Language_Processing/"
    "nlp_sentiment_analysis/data/raw/Data.zip"
)

# Define a separate directory for the extracted raw files.
extraction_path = dataset_zip_path.parent / "Data"

# Create the extraction directory if it does not already exist.
extraction_path.mkdir(parents=True, exist_ok=True)

# Extract the archive contents without modifying the original ZIP file.
with ZipFile(dataset_zip_path, "r") as dataset_archive:
    dataset_archive.extractall(extraction_path)

# Confirm the extraction location.
print(f"Original archive preserved at:\n{dataset_zip_path}")
print(f"\nDataset files extracted to:\n{extraction_path}")

Original archive preserved at:
/content/INF-8239-Ciencia-de-Datos-II/Unit_02_Natural_Language_Processing/nlp_sentiment_analysis/data/raw/Data.zip

Dataset files extracted to:
/content/INF-8239-Ciencia-de-Datos-II/Unit_02_Natural_Language_Processing/nlp_sentiment_analysis/data/raw/Data


In [ ]:
# ==============================================================================
# CELL 11 — LIST THE EXTRACTED DATASET FILES
# Objective: Display every extracted file, its relative path, and its size
# before selecting the file that will be audited.
# ==============================================================================

# Retrieve every extracted file recursively.
extracted_files = sorted(
    file_path
    for file_path in extraction_path.rglob("*")
    if file_path.is_file()
)

# Display the total number of extracted files.
print(f"Total extracted files: {len(extracted_files)}")
print("\nExtracted dataset files:\n")

# Display each file using a path relative to the extraction directory.
for file_path in extracted_files:
    relative_path = file_path.relative_to(extraction_path)
    print(f"{relative_path} ({file_path.stat().st_size:,} bytes)")

Total extracted files: 8

Extracted dataset files:

Corpus_gral/readme.txt (1,667 bytes)
Corpus_gral/reviews_cod.txt (70,105,325 bytes)
Corpus_gral/reviews_tw_definitivo.txt (621,270 bytes)
Corpus_gral/words_cod.txt (1,012,295 bytes)
Datos_all/readme.txt (1,671 bytes)
Datos_all/reviews_cod_all.txt (2,597,659 bytes)
Datos_all/reviews_tw_all.txt (33,000 bytes)
Datos_all/words_cod_all.txt (141,399 bytes)


In [16]:
# ==============================================================================
# CELL 12 — RUN THE DATASET AUDIT
# Objective: Execute the dataset audit using the source modules and
# configuration stored within the project.
# ==============================================================================

# Set the project root as the current working directory.
%cd /content/INF-8239-Ciencia-de-Datos-II/Unit_02_Natural_Language_Processing/nlp_sentiment_analysis

# Execute the audit script as a project module.
!python -m scripts.audit_data

/content/INF-8239-Ciencia-de-Datos-II/Unit_02_Natural_Language_Processing/nlp_sentiment_analysis
DATASET FILE
Path: /content/INF-8239-Ciencia-de-Datos-II/Unit_02_Natural_Language_Processing/nlp_sentiment_analysis/data/processed/spanish_tweets_three_class.csv
SHA-256: c3b250fd915c11f9bf3ac88f5f8d806125d432492e1dbae29c82e481ead84444

DATASET STRUCTURE
Rows: ((40045, 2),)
Columns: ((40045, 2),)
Column names: ['text', 'label']

MISSING AND EMPTY VALUES
text     0
label    0

Empty or whitespace-only texts: 0

DUPLICATES
Duplicated text occurrences: 357

CLASS COUNTS
label
positive    22340
negative    16219
neutral      1486

CLASS PROPORTIONS
label
positive    0.5579
negative    0.4050
neutral     0.0371

TEXT LENGTHS IN CHARACTERS
count    40045.000000
mean        91.066201
std         29.634730
min          2.000000
50%         95.000000
90%        126.000000
99%        140.000000
max        315.000000

AUDIT DECISION
The dataset satisfies the programmed minimum requirements for model t

In [ ]:
# ==============================================================================
# CELL 13 — INSPECT THE EXTRACTED CORPUS FORMAT
# Objective: Examine the documentation and first lines of each extracted file
# to identify the text records, sentiment labels, delimiters, and file structure.
# ==============================================================================

# Import Path for operating-system-independent path management.
from pathlib import Path


# ------------------------------------------------------------------------------
# 1. DEFINE THE EXTRACTED CORPUS DIRECTORY
# ------------------------------------------------------------------------------

# Define the project root directory.
PROJECT_PATH = Path(
    "/content/INF-8239-Ciencia-de-Datos-II/"
    "Unit_02_Natural_Language_Processing/"
    "nlp_sentiment_analysis"
)

# Define the directory containing the extracted public corpus.
EXTRACTED_DATA_PATH = PROJECT_PATH / "data" / "raw" / "Data"

# Verify that the extracted corpus directory exists.
if not EXTRACTED_DATA_PATH.exists():
    raise FileNotFoundError(
        f"The extracted corpus directory was not found:\n"
        f"{EXTRACTED_DATA_PATH}"
    )


# ------------------------------------------------------------------------------
# 2. DEFINE THE FILES THAT REQUIRE INSPECTION
# ------------------------------------------------------------------------------

# Define the extracted text files in a logical inspection order.
files_to_inspect = [
    EXTRACTED_DATA_PATH / "Corpus_gral" / "readme.txt",
    EXTRACTED_DATA_PATH / "Corpus_gral" / "reviews_tw_definitivo.txt",
    EXTRACTED_DATA_PATH / "Corpus_gral" / "reviews_cod.txt",
    EXTRACTED_DATA_PATH / "Corpus_gral" / "words_cod.txt",
    EXTRACTED_DATA_PATH / "Datos_all" / "readme.txt",
    EXTRACTED_DATA_PATH / "Datos_all" / "reviews_tw_all.txt",
    EXTRACTED_DATA_PATH / "Datos_all" / "reviews_cod_all.txt",
    EXTRACTED_DATA_PATH / "Datos_all" / "words_cod_all.txt"
]


# ------------------------------------------------------------------------------
# 3. DISPLAY THE FIRST LINES OF EACH FILE
# ------------------------------------------------------------------------------

# Define how many lines will be displayed from each file.
preview_line_count = 8

# Inspect each file independently.
for file_path in files_to_inspect:

    print("\n" + "=" * 80)
    print(f"FILE: {file_path.relative_to(EXTRACTED_DATA_PATH)}")
    print(f"SIZE: {file_path.stat().st_size:,} bytes")
    print("=" * 80)

    # Skip missing files without stopping the entire inspection.
    if not file_path.exists():
        print("File not found.")
        continue

    # Try common encodings used in Spanish-language corpora.
    file_content = None
    successful_encoding = None

    for encoding_name in ["utf-8", "utf-8-sig", "latin-1", "cp1252"]:

        try:
            file_content = file_path.read_text(
                encoding=encoding_name
            )
            successful_encoding = encoding_name
            break

        except UnicodeDecodeError:
            continue

    # Stop inspecting the current file if no supported encoding works.
    if file_content is None:
        print("The file could not be decoded using the tested encodings.")
        continue

    print(f"Detected encoding: {successful_encoding}")
    print("\nFirst lines:\n")

    # Split the content into lines and display a controlled preview.
    file_lines = file_content.splitlines()

    for line_number, line_content in enumerate(
        file_lines[:preview_line_count],
        start=1
    ):
        print(f"{line_number:02d}: {repr(line_content)}")

    # Display the total number of physical lines in the file.
    print(f"\nTotal physical lines: {len(file_lines):,}")


FILE: Corpus_gral/readme.txt
SIZE: 1,667 bytes
Detected encoding: utf-8

First lines:

01: 'FILE reviews_cod_all.txt '
02: ''
03: 'The file contains the set of reviews. '
04: 'The separator between reviews is the line “%%”.'
05: 'Each review is defined by a sequence of words. Each word is defined in a row.'
06: 'Each word is encoded as a sequence of feature identifiers: '
07: '"id_Dict" "Next1" "Next2" "Next3" "Next4" "Next5" "Back1" "Back2" "Back3" "Back4" "Back5" "Frec1" "Frec2"'
08: ''

Total physical lines: 40

FILE: Corpus_gral/reviews_tw_definitivo.txt
SIZE: 621,270 bytes
Detected encoding: utf-8

First lines:

01: '0;3;5'
02: '1;3;5'
03: '2;3;5'
04: '3;3;5'
05: '4;3;5'
06: '5;3;5'
07: '6;3;5'
08: '7;3;5'

Total physical lines: 63,238

FILE: Corpus_gral/reviews_cod.txt
SIZE: 70,105,325 bytes
Detected encoding: utf-8

First lines:

01: '53386;29964;32816;35047;48483;38726;0;0;0;0;0;0;0'
02: '29964;32816;35047;48483;38726;16251;53386;0;0;0;0;1;1'
03: '32816;35047;48483;38726;16251

In [ ]:
# ==============================================================================
# CELL 14 — READ THE COMPLETE CORPUS DOCUMENTATION
# Objective: Display the complete internal README files to identify the role,
# structure, labels, delimiters, and relationships of the extracted files.
# ==============================================================================

# Import Path for operating-system-independent path management.
from pathlib import Path


# ------------------------------------------------------------------------------
# 1. DEFINE THE EXTRACTED CORPUS DIRECTORY
# ------------------------------------------------------------------------------

# Define the project root directory.
PROJECT_PATH = Path(
    "/content/INF-8239-Ciencia-de-Datos-II/"
    "Unit_02_Natural_Language_Processing/"
    "nlp_sentiment_analysis"
)

# Define the directory containing the extracted public corpus.
EXTRACTED_DATA_PATH = PROJECT_PATH / "data" / "raw" / "Data"


# ------------------------------------------------------------------------------
# 2. DEFINE THE DOCUMENTATION FILES
# ------------------------------------------------------------------------------

# Define the internal documentation files provided with the corpus.
readme_files = [
    EXTRACTED_DATA_PATH / "Corpus_gral" / "readme.txt",
    EXTRACTED_DATA_PATH / "Datos_all" / "readme.txt"
]


# ------------------------------------------------------------------------------
# 3. DISPLAY THE COMPLETE CONTENT OF EACH README FILE
# ------------------------------------------------------------------------------

# Read and display each documentation file independently.
for readme_path in readme_files:

    print("\n" + "=" * 80)
    print(f"FILE: {readme_path.relative_to(EXTRACTED_DATA_PATH)}")
    print("=" * 80)

    # Verify that the documentation file exists.
    if not readme_path.exists():
        print("Documentation file not found.")
        continue

    # Read the complete documentation file.
    readme_content = readme_path.read_text(
        encoding="utf-8"
    )

    # Split the documentation into individual lines.
    readme_lines = readme_content.splitlines()

    # Display every line with its corresponding line number.
    for line_number, line_content in enumerate(
        readme_lines,
        start=1
    ):
        print(f"{line_number:02d}: {line_content}")


FILE: Corpus_gral/readme.txt
01: FILE reviews_cod_all.txt 
02: 
03: The file contains the set of reviews. 
04: The separator between reviews is the line “%%”.
05: Each review is defined by a sequence of words. Each word is defined in a row.
06: Each word is encoded as a sequence of feature identifiers: 
07: "id_Dict" "Next1" "Next2" "Next3" "Next4" "Next5" "Back1" "Back2" "Back3" "Back4" "Back5" "Frec1" "Frec2"
08: 
09: id_Dict: identifier of the current word (dictionary ID).
10: Next1: identifier of the next word in the review (shift = +1).
11: Next2: identifier of the word following Next1 (shift = +2 relative to id_Dict).
12: Next3: identifier of the word following Next2 (shift = +3 relative to id_Dict).
13: Next4: identifier of the word following Next3 (shift = +4 relative to id_Dict).
14: Next5: identifier of the word following Next4 (shift = +5 relative to id_Dict).
15: Back1: identifier of the previous word in the review (shift = −1).
16: Back2: identifier of the word preceding 

In [ ]:
# ==============================================================================
# CELL 15 — VERIFY THE METADATA FIELD STRUCTURE
# Objective: Inspect the second and third metadata fields to determine
# which field contains the documented sentiment labels.
# ==============================================================================

# Import pandas for loading and analyzing the metadata file.
import pandas as pd

# Load the metadata file without assuming column names.
metadata_preview_df = pd.read_csv(
    METADATA_PATH,
    sep=";",
    header=None,
    names=["field_1", "field_2", "field_3"]
)

# Display the metadata dimensions.
print(f"Metadata dimensions: {metadata_preview_df.shape}")

# Display the first ten metadata observations.
print("\nFirst ten metadata rows:")
display(metadata_preview_df.head(10))

# Display the unique values found in the second field.
print("\nUnique values in field_2:")
print(
    sorted(
        metadata_preview_df["field_2"]
        .dropna()
        .unique()
        .tolist()
    )
)

# Display the unique values found in the third field.
print("\nFirst unique values in field_3:")
print(
    sorted(
        metadata_preview_df["field_3"]
        .dropna()
        .unique()
        .tolist()
    )[:30]
)

# Display the frequency distribution of the second field.
print("\nFrequency distribution of field_2:")
print(
    metadata_preview_df["field_2"]
    .value_counts()
    .sort_index()
)

# Display descriptive statistics for the third field.
print("\nDescriptive statistics for field_3:")
print(
    metadata_preview_df["field_3"]
    .describe()
)

Metadata dimensions: (63238, 3)

First ten metadata rows:


,field_1,field_2,field_3
0,0,3,5
1,1,3,5
2,2,3,5
3,3,3,5
4,4,3,5
5,5,3,5
6,6,3,5
7,7,3,5
8,8,3,5
9,9,3,5



Unique values in field_2:
[1, 2, 3, 5]

First unique values in field_3:
[1, 2, 3, 4, 5]

Frequency distribution of field_2:
field_2
1     1008
2      504
3     1899
5    59827
Name: count, dtype: int64

Descriptive statistics for field_3:
count    63238.000000
mean         2.860084
std          1.219197
min          1.000000
25%          1.000000
50%          3.000000
75%          4.000000
max          5.000000
Name: field_3, dtype: float64


In [ ]:
# ==============================================================================
# CELL 16 — INVESTIGATE THE UNDOCUMENTED LABEL VALUE
# Objective: Analyze the relationship between dataset partitions and label
# values before defining the final sentiment mapping.
# ==============================================================================

# Import pandas for tabular analysis.
import pandas as pd

# Assign descriptive names to the three metadata fields.
metadata_analysis_df = metadata_preview_df.rename(
    columns={
        "field_1": "identifier",
        "field_2": "dataset_set",
        "field_3": "label_code"
    }
)

# Display the absolute frequency of each label code.
print("=" * 70)
print("LABEL CODE DISTRIBUTION")
print("=" * 70)
print(
    metadata_analysis_df["label_code"]
    .value_counts()
    .sort_index()
)

# Display the percentage distribution of each label code.
print("\n" + "=" * 70)
print("LABEL CODE PERCENTAGES")
print("=" * 70)
print(
    metadata_analysis_df["label_code"]
    .value_counts(normalize=True)
    .sort_index()
    .mul(100)
    .round(2)
)

# Create a cross-tabulation between dataset partitions and label codes.
set_label_table = pd.crosstab(
    metadata_analysis_df["dataset_set"],
    metadata_analysis_df["label_code"],
    margins=True
)

print("\n" + "=" * 70)
print("DATASET SET BY LABEL CODE")
print("=" * 70)
display(set_label_table)

# Display sample records associated with the undocumented label code 5.
print("\n" + "=" * 70)
print("SAMPLE RECORDS WITH LABEL CODE 5")
print("=" * 70)
display(
    metadata_analysis_df[
        metadata_analysis_df["label_code"] == 5
    ].head(20)
)

LABEL CODE DISTRIBUTION
label_code
1    16219
2     1486
3    22356
4    21278
5     1899
Name: count, dtype: int64

LABEL CODE PERCENTAGES
label_code
1    25.65
2     2.35
3    35.35
4    33.65
5     3.00
Name: proportion, dtype: float64

DATASET SET BY LABEL CODE


label_code,1,2,3,4,5,All
dataset_set,,,,,,
1,418,133,318,139,0,1008
2,219,68,156,61,0,504
3,0,0,0,0,1899,1899
5,15582,1285,21882,21078,0,59827
All,16219,1486,22356,21278,1899,63238



SAMPLE RECORDS WITH LABEL CODE 5


,identifier,dataset_set,label_code
0,0,3,5
1,1,3,5
2,2,3,5
3,3,3,5
4,4,3,5
5,5,3,5
6,6,3,5
7,7,3,5
8,8,3,5
9,9,3,5


In [ ]:
# ==============================================================================
# CELL 17 — RECONSTRUCT AND AUDIT THE DOWNLOADED PUBLIC CORPUS
# Objective: Reconstruct the original texts, preserve all documented records,
# distinguish unlabeled evaluation data, and audit the downloaded corpus.
# ==============================================================================

# Import hashlib to calculate SHA-256 file identifiers.
import hashlib

# Import Path for operating-system-independent path management.
from pathlib import Path

# Import pandas for creating and auditing the reconstructed dataset.
import pandas as pd


# ------------------------------------------------------------------------------
# 1. DEFINE THE CORPUS FILES
# ------------------------------------------------------------------------------

# Define the project root directory.
PROJECT_PATH = Path(
    "/content/INF-8239-Ciencia-de-Datos-II/"
    "Unit_02_Natural_Language_Processing/"
    "nlp_sentiment_analysis"
)

# Define the General Corpus directory.
CORPUS_PATH = PROJECT_PATH / "data" / "raw" / "Data" / "Corpus_gral"

# Define the original downloaded archive.
ZIP_PATH = PROJECT_PATH / "data" / "raw" / "Data.zip"

# Define the files required to reconstruct the corpus.
METADATA_PATH = CORPUS_PATH / "reviews_tw_definitivo.txt"
ENCODED_REVIEWS_PATH = CORPUS_PATH / "reviews_cod.txt"
DICTIONARY_PATH = CORPUS_PATH / "words_cod.txt"

# Verify that every required file exists.
required_files = [
    ZIP_PATH,
    METADATA_PATH,
    ENCODED_REVIEWS_PATH,
    DICTIONARY_PATH
]

for file_path in required_files:
    if not file_path.exists():
        raise FileNotFoundError(
            f"Required corpus file not found:\n{file_path}"
        )


# ------------------------------------------------------------------------------
# 2. DEFINE THE SHA-256 FUNCTION
# ------------------------------------------------------------------------------

def calculate_sha256(file_path: Path) -> str:
    """Calculate the SHA-256 hash of a file."""

    # Create a SHA-256 hashing object.
    sha256_hash = hashlib.sha256()

    # Read the file in binary chunks.
    with file_path.open("rb") as file_object:
        for chunk in iter(lambda: file_object.read(8192), b""):
            sha256_hash.update(chunk)

    # Return the final hexadecimal identifier.
    return sha256_hash.hexdigest()


# ------------------------------------------------------------------------------
# 3. LOAD THE WORD DICTIONARY
# ------------------------------------------------------------------------------

# Create a mapping from word identifiers to original words.
id_to_word = {}

# Read the word dictionary.
with DICTIONARY_PATH.open("r", encoding="utf-8") as dictionary_file:

    for line in dictionary_file:
        # Remove surrounding whitespace.
        line = line.strip()

        # Ignore empty or malformed lines.
        if not line or ";" not in line:
            continue

        # Separate the word from its identifier.
        word, word_identifier = line.rsplit(";", 1)

        # Store the dictionary relationship.
        id_to_word[int(word_identifier)] = word

print(f"Dictionary entries loaded: {len(id_to_word):,}")


# ------------------------------------------------------------------------------
# 4. LOAD THE REVIEW METADATA
# ------------------------------------------------------------------------------

# Define the dataset-set mapping identified from the documentation and files.
set_mapping = {
    1: "training",
    2: "test",
    3: "evaluation",
    5: "general"
}

# Define the sentiment-label mapping.
# Code 5 is not a sentiment class; it identifies unavailable evaluation labels.
label_mapping = {
    1: "negative",
    2: "neutral",
    3: "positive",
    4: "none",
    5: "unlabeled"
}

# Load the metadata file.
metadata_df = pd.read_csv(
    METADATA_PATH,
    sep=";",
    header=None,
    names=[
        "identifier",
        "set_code",
        "label_code"
    ]
)

# Convert the documented codes into readable categories.
metadata_df["dataset_set"] = metadata_df["set_code"].map(set_mapping)
metadata_df["label"] = metadata_df["label_code"].map(label_mapping)

# Verify that every code was mapped successfully.
if metadata_df["dataset_set"].isna().any():
    raise ValueError(
        "One or more dataset-set codes are not documented."
    )

if metadata_df["label"].isna().any():
    raise ValueError(
        "One or more label codes are not documented."
    )

print(f"Metadata records loaded: {len(metadata_df):,}")


# ------------------------------------------------------------------------------
# 5. RECONSTRUCT THE TEXTS
# ------------------------------------------------------------------------------

# Create a list for the reconstructed texts.
reconstructed_texts = []

# Store the words of the current review.
current_review_words = []

# Count word identifiers not found in the dictionary.
unknown_word_identifiers = 0

# Read the encoded corpus.
with ENCODED_REVIEWS_PATH.open(
    "r",
    encoding="utf-8"
) as encoded_reviews_file:

    for line in encoded_reviews_file:
        # Remove surrounding whitespace.
        line = line.strip()

        # The %% line marks the end of a review.
        if line == "%%":
            reconstructed_texts.append(
                " ".join(current_review_words).strip()
            )

            current_review_words = []
            continue

        # Ignore empty lines.
        if not line:
            continue

        # The first field contains the current word identifier.
        current_word_identifier = int(
            line.split(";", 1)[0]
        )

        # Retrieve the original word.
        current_word = id_to_word.get(
            current_word_identifier
        )

        # Add known words to the current review.
        if current_word is not None:
            current_review_words.append(current_word)
        else:
            unknown_word_identifiers += 1

# Preserve the final review if the file does not end with %%.
if current_review_words:
    reconstructed_texts.append(
        " ".join(current_review_words).strip()
    )

print(f"Reviews reconstructed: {len(reconstructed_texts):,}")
print(f"Unknown word identifiers: {unknown_word_identifiers:,}")


# ------------------------------------------------------------------------------
# 6. VERIFY THE TEXT-METADATA RELATIONSHIP
# ------------------------------------------------------------------------------

# Confirm that every metadata record has one reconstructed text.
if len(reconstructed_texts) != len(metadata_df):
    raise ValueError(
        "The number of reconstructed reviews does not match the metadata.\n"
        f"Reconstructed reviews: {len(reconstructed_texts):,}\n"
        f"Metadata records:      {len(metadata_df):,}"
    )

# Add the reconstructed texts to the metadata.
corpus_df = metadata_df.copy()
corpus_df["text"] = reconstructed_texts

# Organize the dataset columns.
corpus_df = corpus_df[
    [
        "identifier",
        "set_code",
        "dataset_set",
        "text",
        "label_code",
        "label"
    ]
]


# ------------------------------------------------------------------------------
# 7. REPORT FILE IDENTITY AND STRUCTURE
# ------------------------------------------------------------------------------

print("\n" + "=" * 70)
print("FILE AND DATASET STRUCTURE")
print("=" * 70)
print(f"Original archive: {ZIP_PATH}")
print(f"Archive SHA-256:  {calculate_sha256(ZIP_PATH)}")
print(f"Rows:             {corpus_df.shape,}")
print(f"Columns:          {corpus_df.shape,}")

print("\nColumn names:")

for column_name in corpus_df.columns:
    print(f"- {column_name}")


# ------------------------------------------------------------------------------
# 8. REPORT MISSING AND EMPTY VALUES
# ------------------------------------------------------------------------------

# Calculate missing values.
missing_values = corpus_df.isna().sum()

# Calculate empty reconstructed texts.
empty_text_count = (
    corpus_df["text"]
    .fillna("")
    .astype(str)
    .str.strip()
    .eq("")
    .sum()
)

print("\n" + "=" * 70)
print("MISSING AND EMPTY VALUES")
print("=" * 70)
print(missing_values.to_string())
print(f"\nEmpty or whitespace-only texts: {empty_text_count:,}")


# ------------------------------------------------------------------------------
# 9. REPORT DUPLICATES
# ------------------------------------------------------------------------------

# Calculate complete duplicate rows.
duplicate_rows = corpus_df.duplicated().sum()

# Calculate additional duplicated texts.
duplicate_texts = corpus_df.duplicated(
    subset=["text"]
).sum()

# Calculate additional duplicated text-label pairs.
duplicate_text_label_pairs = corpus_df.duplicated(
    subset=["text", "label"]
).sum()

print("\n" + "=" * 70)
print("DUPLICATES")
print("=" * 70)
print(f"Duplicate complete rows:       {duplicate_rows:,}")
print(f"Duplicate text occurrences:    {duplicate_texts:,}")
print(f"Duplicate text-label pairs:    {duplicate_text_label_pairs:,}")


# ------------------------------------------------------------------------------
# 10. REPORT PARTITION DISTRIBUTION
# ------------------------------------------------------------------------------

# Calculate the number of observations in each dataset partition.
set_distribution = pd.crosstab(
    corpus_df["dataset_set"],
    corpus_df["label"],
    margins=True
)

print("\n" + "=" * 70)
print("DATASET SET AND LABEL DISTRIBUTION")
print("=" * 70)
display(set_distribution)


# ------------------------------------------------------------------------------
# 11. REPORT THE COMPLETE LABEL DISTRIBUTION
# ------------------------------------------------------------------------------

# Calculate complete counts and percentages, including unlabeled observations.
all_label_counts = corpus_df["label"].value_counts(
    dropna=False
)

all_label_percentages = (
    corpus_df["label"]
    .value_counts(normalize=True, dropna=False)
    .mul(100)
)

all_label_distribution = pd.DataFrame({
    "count": all_label_counts,
    "percentage": all_label_percentages
})

print("\n" + "=" * 70)
print("COMPLETE LABEL DISTRIBUTION")
print("=" * 70)
display(all_label_distribution.round(2))


# ------------------------------------------------------------------------------
# 12. REPORT THE SUPERVISED LABEL DISTRIBUTION
# ------------------------------------------------------------------------------

# Exclude records whose sentiment label is unavailable.
labeled_corpus_df = corpus_df[
    corpus_df["label"] != "unlabeled"
].copy()

# Calculate the distribution of documented sentiment labels.
supervised_label_counts = (
    labeled_corpus_df["label"]
    .value_counts()
)

supervised_label_percentages = (
    labeled_corpus_df["label"]
    .value_counts(normalize=True)
    .mul(100)
)

supervised_label_distribution = pd.DataFrame({
    "count": supervised_label_counts,
    "percentage": supervised_label_percentages
})

print("\n" + "=" * 70)
print("SUPERVISED LABEL DISTRIBUTION")
print("=" * 70)
display(supervised_label_distribution.round(2))


# ------------------------------------------------------------------------------
# 13. REPORT THE THREE-CLASS DISTRIBUTION
# ------------------------------------------------------------------------------

# Keep only the target classes proposed for the future model.
three_class_df = corpus_df[
    corpus_df["label"].isin(
        ["negative", "neutral", "positive"]
    )
].copy()

# Calculate the three-class distribution.
three_class_counts = three_class_df["label"].value_counts()

three_class_percentages = (
    three_class_df["label"]
    .value_counts(normalize=True)
    .mul(100)
)

three_class_distribution = pd.DataFrame({
    "count": three_class_counts,
    "percentage": three_class_percentages
})

print("\n" + "=" * 70)
print("PROPOSED THREE-CLASS DISTRIBUTION")
print("=" * 70)
display(three_class_distribution.round(2))


# ------------------------------------------------------------------------------
# 14. REPORT TEXT-LENGTH STATISTICS
# ------------------------------------------------------------------------------

# Calculate text lengths in characters.
text_lengths = corpus_df["text"].str.len()

# Calculate the requested percentiles.
text_length_summary = text_lengths.describe(
    percentiles=[0.50, 0.90, 0.99]
)

print("\n" + "=" * 70)
print("TEXT LENGTHS IN CHARACTERS")
print("=" * 70)
print(text_length_summary.to_string())


# ------------------------------------------------------------------------------
# 15. DISPLAY RECONSTRUCTED EXAMPLES
# ------------------------------------------------------------------------------

print("\n" + "=" * 70)
print("RECONSTRUCTED DATA SAMPLE")
print("=" * 70)

# Display examples from the reconstructed corpus.
display(
    corpus_df[
        [
            "identifier",
            "dataset_set",
            "text",
            "label"
        ]
    ].head()
)


# ------------------------------------------------------------------------------
# 16. COMPLETE THE AUDIT
# ------------------------------------------------------------------------------

print("\n" + "=" * 70)
print("DOWNLOADED PUBLIC CORPUS AUDIT COMPLETED")
print("=" * 70)

Dictionary entries loaded: 53,385
Metadata records loaded: 63,238
Reviews reconstructed: 63,238
Unknown word identifiers: 203,114

FILE AND DATASET STRUCTURE
Original archive: /content/INF-8239-Ciencia-de-Datos-II/Unit_02_Natural_Language_Processing/nlp_sentiment_analysis/data/raw/Data.zip
Archive SHA-256:  42709113f5b10ef421ad0b900f6b836f28fa02073ca3d0869edb689bc9c825d2
Rows:             ((63238, 6),)
Columns:          ((63238, 6),)

Column names:
- identifier
- set_code
- dataset_set
- text
- label_code
- label

MISSING AND EMPTY VALUES
identifier     0
set_code       0
dataset_set    0
text           0
label_code     0
label          0

Empty or whitespace-only texts: 114

DUPLICATES
Duplicate complete rows:       0
Duplicate text occurrences:    1,599
Duplicate text-label pairs:    1,565

DATASET SET AND LABEL DISTRIBUTION


label,negative,neutral,none,positive,unlabeled,All
dataset_set,,,,,,
evaluation,0,0,0,0,1899,1899
general,15582,1285,21078,21882,0,59827
test,219,68,61,156,0,504
training,418,133,139,318,0,1008
All,16219,1486,21278,22356,1899,63238



COMPLETE LABEL DISTRIBUTION


,count,percentage
label,,
positive,22356,35.35
none,21278,33.65
negative,16219,25.65
unlabeled,1899,3.00
neutral,1486,2.35



SUPERVISED LABEL DISTRIBUTION


,count,percentage
label,,
positive,22356,36.45
none,21278,34.69
negative,16219,26.44
neutral,1486,2.42



PROPOSED THREE-CLASS DISTRIBUTION


,count,percentage
label,,
positive,22356,55.80
negative,16219,40.49
neutral,1486,3.71



TEXT LENGTHS IN CHARACTERS
count    63238.000000
mean        84.285904
std         32.933300
min          0.000000
50%         89.000000
90%        124.000000
99%        139.000000
max        315.000000

RECONSTRUCTED DATA SAMPLE


,identifier,dataset_set,text,label
0,0,evaluation,lonelysoad mientras no te pillar el uno semana...,unlabeled
1,1,evaluation,ceemeese ya ser hora de volver el csgo dejar e...,unlabeled
2,2,evaluation,mireiaescribano justo cuando se terminar el fi...,unlabeled
3,3,evaluation,pensba ir hacer wxplosive uno reflexión profun...,unlabeled
4,4,evaluation,vic_phantomhive_si el encuentro sin compañero ...,unlabeled



DOWNLOADED PUBLIC CORPUS AUDIT COMPLETED


## Downloaded Corpus Audit Results

The downloaded corpus was reconstructed from its encoded representation using the provided word dictionary, review separators, metadata, and sentiment-label codes.

A total of **63,238 records** were reconstructed, matching the number of metadata observations. No missing values were detected in the reconstructed columns. However, **114 records contained empty or whitespace-only text** and should be excluded from the processed modeling dataset.

The audit identified **1,599 duplicated text occurrences** and **1,565 duplicated text-label pairs**. These duplicates must be controlled before creating or modifying model partitions because duplicated observations across training and test sets could artificially inflate model performance.

The complete corpus contains 22,356 positive records, 16,219 negative records, 1,486 neutral records, 21,278 records labeled as no opinion, and 1,899 evaluation records without disclosed sentiment labels.

After preserving only the proposed negative, neutral, and positive classes, the dataset contains **40,061 labeled observations**. Positive represents 55.80%, negative represents 40.49%, and neutral represents only 3.71%. The severe underrepresentation of neutral sentiment is the principal class-quality limitation.

The reconstruction also encountered **203,114 unknown word-identifier occurrences** that could not be recovered using the provided dictionary. As a result, the reconstructed text may omit some original tokens. This limitation must be documented in the dataset card.

Text lengths range from 0 to 315 characters. The median length is 89 characters, while the 90th and 99th percentiles are 124 and 139 characters, respectively.

The corpus is approved with conditions. Empty texts, records without disclosed labels, the no-opinion category, and duplicate texts must be treated explicitly before modeling.

In [ ]:
# ==============================================================================
# CELL 18 — ANALYZE UNKNOWN WORD IDENTIFIERS
# Objective: Identify the most frequent encoded word identifiers that could
# not be recovered from the provided corpus dictionary.
# ==============================================================================

# Import Counter to calculate unknown identifier frequencies.
from collections import Counter

# Create a counter for unknown word identifiers.
unknown_identifier_counts = Counter()

# Read the encoded reviews again.
with ENCODED_REVIEWS_PATH.open(
    "r",
    encoding="utf-8"
) as encoded_reviews_file:

    for line in encoded_reviews_file:
        # Remove surrounding whitespace.
        line = line.strip()

        # Ignore review separators and empty lines.
        if not line or line == "%%":
            continue

        # Retrieve the current word identifier.
        current_word_identifier = int(
            line.split(";", 1)[0]
        )

        # Count identifiers not available in the dictionary.
        if current_word_identifier not in id_to_word:
            unknown_identifier_counts[current_word_identifier] += 1

# Display the number of distinct unknown identifiers.
print(
    "Distinct unknown word identifiers:",
    len(unknown_identifier_counts)
)

# Create a table with the most frequent unknown identifiers.
unknown_identifiers_df = pd.DataFrame(
    unknown_identifier_counts.most_common(20),
    columns=[
        "word_identifier",
        "occurrences"
    ]
)

# Display the most frequent unknown identifiers.
display(unknown_identifiers_df)

Distinct unknown word identifiers: 1


,word_identifier,occurrences
0,53386,203114


## Text Reconstruction

The downloaded corpus stores each review as a sequence of encoded word
identifiers. The original text was reconstructed by mapping each `id_Dict`
value to the corresponding word in the supplied dictionary and joining the
tokens until the `%%` review separator was reached.

The audit detected 203,114 occurrences of identifiers not present in the
dictionary. Further analysis showed that every occurrence corresponded to
the same identifier: `0`.

Identifier `0` was interpreted as a reserved or padding value rather than a
recoverable lexical token. It was therefore excluded during text
reconstruction. All positive word identifiers were successfully resolved
using the supplied dictionary.

This means that the audit did not identify 203,114 different unknown words.
It identified one reserved code repeated 203,114 times.

In [ ]:
# ==============================================================================
# CELL 19 — SAVE THE APPROVED THREE-CLASS DATASET
# Objective: Create a processed CSV containing non-empty texts with the
# approved negative, neutral, and positive sentiment classes.
# ==============================================================================

# Define the processed data directory.
processed_data_path = PROJECT_PATH / "data" / "processed"

# Create the directory if it does not already exist.
processed_data_path.mkdir(parents=True, exist_ok=True)

# Keep only the three approved sentiment classes.
processed_df = corpus_df[
    corpus_df["label"].isin(
        ["negative", "neutral", "positive"]
    )
].copy()

# Remove surrounding whitespace from reconstructed texts.
processed_df["text"] = (
    processed_df["text"]
    .astype(str)
    .str.strip()
)

# Exclude empty or whitespace-only texts.
processed_df = processed_df[
    processed_df["text"] != ""
].copy()

# Select and rename the columns required by the data contract.
processed_df = processed_df[
    ["text", "label"]
].copy()

# Define the processed dataset path.
processed_dataset_path = (
    processed_data_path / "spanish_tweets_three_class.csv"
)

# Save the processed dataset without adding a pandas index column.
processed_df.to_csv(
    processed_dataset_path,
    index=False,
    encoding="utf-8"
)

# Display the saved dataset information.
print(f"Processed dataset saved to:\n{processed_dataset_path}")
print(f"\nRows and columns: {processed_df.shape}")
print("\nClass distribution:")
print(processed_df["label"].value_counts())

Processed dataset saved to:
/content/INF-8239-Ciencia-de-Datos-II/Unit_02_Natural_Language_Processing/nlp_sentiment_analysis/data/processed/spanish_tweets_three_class.csv

Rows and columns: (40045, 2)

Class distribution:
label
positive    22340
negative    16219
neutral      1486
Name: count, dtype: int64


In [ ]:
# ==============================================================================
# CELL 20 — CONFIGURE THE PROCESSED DATASET
# Objective: Update the local project configuration to use the approved
# three-class processed corpus.
# ==============================================================================

# Define the hidden environment configuration file.
env_path = PROJECT_PATH / ".env"

# Define the final dataset configuration.
env_content = """DATA_SOURCE=local
DATASET_PATH=data/processed/spanish_tweets_three_class.csv
DATASET_URL=
TEXT_COLUMN=text
TARGET_COLUMN=label
RANDOM_STATE=42
"""

# Save the final local configuration.
env_path.write_text(
    env_content,
    encoding="utf-8"
)

# Display the saved configuration for verification.
print(env_path.read_text(encoding="utf-8"))

DATA_SOURCE=local
DATASET_PATH=data/processed/spanish_tweets_three_class.csv
DATASET_URL=
TEXT_COLUMN=text
TARGET_COLUMN=label
RANDOM_STATE=42



In [ ]:
# ==============================================================================
# CELL 21 — EXECUTE THE DATA CONTRACT
# Objective: Run the automated tests that validate the configured dataset
# structure and programmed quality requirements.
# ==============================================================================

# Set the project root as the current working directory.
%cd /content/INF-8239-Ciencia-de-Datos-II/Unit_02_Natural_Language_Processing/nlp_sentiment_analysis

# Execute only the data-contract tests using the current Colab environment.
!python -m pytest test/test_data_contract.py -q

/content/INF-8239-Ciencia-de-Datos-II/Unit_02_Natural_Language_Processing/nlp_sentiment_analysis
...                                                                      [100%]
3 passed in 0.40s


In [ ]:
# ==============================================================================
# CELL 22 — VERIFY LOCAL AND CLOUD PORTABILITY
# Objective: Check the project for absolute paths, protected environment files,
# required dependency files, and large files that should not be committed.
# ==============================================================================

# Import Path for operating-system-independent path management.
from pathlib import Path

# Define the project root directory.
PROJECT_PATH = Path(
    "/content/INF-8239-Ciencia-de-Datos-II/"
    "Unit_02_Natural_Language_Processing/"
    "nlp_sentiment_analysis"
)

# Define the files required for local and Colab reproducibility.
required_portability_files = [
    PROJECT_PATH / "pyproject.toml",
    PROJECT_PATH / "uv.lock",
    PROJECT_PATH / "requirements-colab.txt",
    PROJECT_PATH / ".env.example",
    PROJECT_PATH / ".gitignore",
    PROJECT_PATH / "README.md"
]

# Display whether each required portability file exists.
print("=" * 70)
print("REQUIRED PORTABILITY FILES")
print("=" * 70)

for file_path in required_portability_files:
    status = "FOUND" if file_path.exists() else "MISSING"
    print(f"[{status}] {file_path.name}")

# Define text-based files that can be inspected for absolute paths.
inspectable_extensions = {
    ".py",
    ".md",
    ".toml",
    ".txt",
    ".yml",
    ".yaml"
}

# Search for files containing environment-specific absolute paths.
absolute_path_matches = []

for file_path in PROJECT_PATH.rglob("*"):
    if not file_path.is_file():
        continue

    if file_path.suffix.lower() not in inspectable_extensions:
        continue

    try:
        file_content = file_path.read_text(
            encoding="utf-8",
            errors="ignore"
        )
    except OSError:
        continue

    # Detect common Colab and Windows absolute-path patterns.
    if "/content/" in file_content or "C:\\Users\\" in file_content:
        absolute_path_matches.append(
            file_path.relative_to(PROJECT_PATH)
        )

print("\n" + "=" * 70)
print("ABSOLUTE PATH CHECK")
print("=" * 70)

if absolute_path_matches:
    print("Files containing possible absolute paths:")

    for file_path in absolute_path_matches:
        print(f"- {file_path}")
else:
    print("No absolute paths detected in inspected project files.")

# Identify large files using a 10 MB review threshold.
large_files = []

for file_path in PROJECT_PATH.rglob("*"):
    if file_path.is_file() and file_path.stat().st_size > 10 * 1024 * 1024:
        large_files.append(
            (
                file_path.relative_to(PROJECT_PATH),
                file_path.stat().st_size
            )
        )

print("\n" + "=" * 70)
print("LARGE FILE CHECK")
print("=" * 70)

if large_files:
    for file_path, file_size in large_files:
        print(
            f"- {file_path}: "
            f"{file_size / (1024 * 1024):.2f} MB"
        )
else:
    print("No files larger than 10 MB were detected.")

REQUIRED PORTABILITY FILES
[MISSING] pyproject.toml
[MISSING] uv.lock
[MISSING] requirements-colab.txt
[MISSING] .env.example
[MISSING] .gitignore
[FOUND] README.md

ABSOLUTE PATH CHECK
No absolute paths detected in inspected project files.

LARGE FILE CHECK
- data/raw/Data.zip: 10.20 MB
- data/raw/Data/Corpus_gral/reviews_cod.txt: 66.86 MB


In [ ]:
# ==============================================================================
# CELL 24 — UPDATE THE REPOSITORY GITIGNORE
# Objective: Preserve the existing Git ignore rules and add rules for local
# environment files, downloaded datasets, and generated processed datasets.
# ==============================================================================

# Import Path for operating-system-independent path management.
from pathlib import Path

# Define the repository root directory.
REPOSITORY_PATH = Path(
    "/content/INF-8239-Ciencia-de-Datos-II"
)

# Define the repository-level .gitignore file.
gitignore_path = REPOSITORY_PATH / ".gitignore"

# Define the rules required for datasets stored inside nested projects.
required_rules = [
    "**/.env",
    "**/data/raw/*",
    "!**/data/raw/.gitkeep",
    "**/data/processed/*",
    "!**/data/processed/.gitkeep"
]

# Read the existing .gitignore content.
existing_content = gitignore_path.read_text(
    encoding="utf-8"
)

# Collect the active rules already present in the file.
existing_rules = {
    line.strip()
    for line in existing_content.splitlines()
    if line.strip() and not line.strip().startswith("#")
}

# Identify only the rules that are not already present.
missing_rules = [
    rule
    for rule in required_rules
    if rule not in existing_rules
]

# Add the missing rules without deleting the existing configuration.
if missing_rules:
    updated_content = existing_content.rstrip()

    updated_content += (
        "\n\n"
        "# Local environment configuration inside any project\n"
    )

    if "**/.env" in missing_rules:
        updated_content += "**/.env\n"

    updated_content += (
        "\n"
        "# Downloaded raw datasets inside any project\n"
    )

    if "**/data/raw/*" in missing_rules:
        updated_content += "**/data/raw/*\n"

    if "!**/data/raw/.gitkeep" in missing_rules:
        updated_content += "!**/data/raw/.gitkeep\n"

    updated_content += (
        "\n"
        "# Generated processed datasets inside any project\n"
    )

    if "**/data/processed/*" in missing_rules:
        updated_content += "**/data/processed/*\n"

    if "!**/data/processed/.gitkeep" in missing_rules:
        updated_content += "!**/data/processed/.gitkeep\n"

    # Save the updated .gitignore file.
    gitignore_path.write_text(
        updated_content,
        encoding="utf-8"
    )

    print("The repository-level .gitignore was updated.")
    print("\nRules added:")

    for rule in missing_rules:
        print(f"- {rule}")

else:
    print("All required rules are already present.")

# Display the final .gitignore content.
print("\n" + "=" * 70)
print("FINAL .GITIGNORE CONTENT")
print("=" * 70)
print(gitignore_path.read_text(encoding="utf-8"))

The repository-level .gitignore was updated.

Rules added:
- **/.env
- **/data/raw/*
- !**/data/raw/.gitkeep
- **/data/processed/*
- !**/data/processed/.gitkeep

FINAL .GITIGNORE CONTENT
.venv/
__pycache__/
.pytest_cache/
.ipynb_checkpoints/
data/raw/*
!data/raw/.gitkeep

# Local environment configuration inside any project
**/.env

# Downloaded raw datasets inside any project
**/data/raw/*
!**/data/raw/.gitkeep

# Generated processed datasets inside any project
**/data/processed/*
!**/data/processed/.gitkeep



In [17]:
# ==============================================================================
# LAB05 — CELL 01: VERIFY THE PROCESSED TEXT DATASET
# Objective: Confirm that the approved three-class corpus is available and
# satisfies the minimum requirements before training the classifiers.
# ==============================================================================

# Import Path for operating-system-independent path management.
from pathlib import Path

# Import pandas for loading and inspecting the processed dataset.
import pandas as pd


# Define the existing NLP project root directory.
PROJECT_PATH = Path(
    "/content/INF-8239-Ciencia-de-Datos-II/"
    "Unit_02_Natural_Language_Processing/"
    "nlp_sentiment_analysis"
)

# Define the processed three-class dataset created during LAB04.
DATASET_PATH = (
    PROJECT_PATH
    / "data"
    / "processed"
    / "spanish_tweets_three_class.csv"
)

# Stop execution if the processed dataset is unavailable.
if not DATASET_PATH.exists():
    raise FileNotFoundError(
        "The processed three-class dataset was not found:\n"
        f"{DATASET_PATH}"
    )

# Load the processed dataset.
dataset_df = pd.read_csv(DATASET_PATH)

# Define the required columns.
TEXT_COLUMN = "text"
TARGET_COLUMN = "label"

# Identify required columns that are missing.
missing_columns = [
    column_name
    for column_name in [TEXT_COLUMN, TARGET_COLUMN]
    if column_name not in dataset_df.columns
]

# Stop execution if a required column is unavailable.
if missing_columns:
    raise ValueError(
        f"Required columns are missing: {missing_columns}"
    )

# Convert text values into stripped strings for structural verification.
clean_text = (
    dataset_df[TEXT_COLUMN]
    .fillna("")
    .astype(str)
    .str.strip()
)

# Count empty or whitespace-only texts.
empty_text_count = clean_text.eq("").sum()

# Count missing target labels.
missing_label_count = dataset_df[TARGET_COLUMN].isna().sum()

# Count duplicated text observations.
duplicate_text_count = dataset_df.duplicated(
    subset=[TEXT_COLUMN]
).sum()

# Retrieve the available target classes.
available_classes = sorted(
    dataset_df[TARGET_COLUMN]
    .dropna()
    .unique()
    .tolist()
)

# Define the expected target classes.
expected_classes = [
    "negative",
    "neutral",
    "positive"
]

# Display the dataset verification results.
print("=" * 70)
print("LAB05 DATASET VERIFICATION")
print("=" * 70)
print(f"Dataset path:        {DATASET_PATH}")
print(f"Rows:                {dataset_df.shape,}")
print(f"Columns:             {dataset_df.shape[1]}")
print(f"Column names:        {dataset_df.columns.tolist()}")
print(f"Empty texts:         {empty_text_count:,}")
print(f"Missing labels:      {missing_label_count:,}")
print(f"Duplicated texts:    {duplicate_text_count:,}")
print(f"Available classes:   {available_classes}")
print(f"Expected classes:    {expected_classes}")

# Verify that the dataset contains the three expected classes.
if available_classes != expected_classes:
    raise ValueError(
        "The available target classes do not match the expected classes."
    )

# Display the absolute and relative class distribution.
class_distribution = pd.DataFrame({
    "count": dataset_df[TARGET_COLUMN].value_counts(),
    "percentage": (
        dataset_df[TARGET_COLUMN]
        .value_counts(normalize=True)
        .mul(100)
        .round(2)
    )
})

print("\nClass distribution:")
display(class_distribution)

# Display a small sample for qualitative verification.
print("\nDataset sample:")
display(
    dataset_df[
        [TEXT_COLUMN, TARGET_COLUMN]
    ].head()
)

# Confirm that the processed corpus is ready for the training stage.
print("\nThe processed corpus is available for LAB05.")

LAB05 DATASET VERIFICATION
Dataset path:        /content/INF-8239-Ciencia-de-Datos-II/Unit_02_Natural_Language_Processing/nlp_sentiment_analysis/data/processed/spanish_tweets_three_class.csv
Rows:                ((40045, 2),)
Columns:             2
Column names:        ['text', 'label']
Empty texts:         0
Missing labels:      0
Duplicated texts:    357
Available classes:   ['negative', 'neutral', 'positive']
Expected classes:    ['negative', 'neutral', 'positive']

Class distribution:


,count,percentage
label,,
positive,22340,55.79
negative,16219,40.50
neutral,1486,3.71



Dataset sample:


,text,label
0,noseashetero de verdad que ir decir petar que ...,positive
1,piscolabisaereo historiang sposteguillo lo ten...,positive
2,el final ir bueno mañana tener fiesta así que ...,positive
3,yo no tener tiempo para ese cosa ahora mismo,negative
4,misschaotic ve ese brillo ser uno coso que hac...,negative



The processed corpus is available for LAB05.


In [18]:
# ==============================================================================
# LAB05 — CELL 02: LOCATE UNIT 02 REQUIREMENTS FILES
# Objective: Locate the available dependency files inside Unit 02 before
# installing any project packages.
# ==============================================================================

# Import Path for operating-system-independent path management.
from pathlib import Path

# Define the Unit 02 directory.
UNIT_02_PATH = Path(
    "/content/INF-8239-Ciencia-de-Datos-II/"
    "Unit_02_Natural_Language_Processing"
)

# Search recursively for requirements files inside Unit 02.
requirements_files = sorted(
    UNIT_02_PATH.rglob("requirements*.txt")
)

# Display every dependency file found.
print("Requirements files found in Unit 02:\n")

if requirements_files:
    for file_number, file_path in enumerate(
        requirements_files,
        start=1
    ):
        print(f"{file_number}. {file_path}")
else:
    print("No requirements files were found inside Unit 02.")

# Set the NLP project as the current working directory.
%cd /content/INF-8239-Ciencia-de-Datos-II/Unit_02_Natural_Language_Processing/nlp_sentiment_analysis

# Install the dependencies from the existing Colab requirements file.
%pip install -q -r requirements.txt

# Install Gensim because Word2Vec is required by embeddings_network.py.
%pip install -q gensim



Requirements files found in Unit 02:

1. /content/INF-8239-Ciencia-de-Datos-II/Unit_02_Natural_Language_Processing/cv_image_classification/requirements.txt
2. /content/INF-8239-Ciencia-de-Datos-II/Unit_02_Natural_Language_Processing/nlp_sentiment_analysis/requirements.txt
/content/INF-8239-Ciencia-de-Datos-II/Unit_02_Natural_Language_Processing/nlp_sentiment_analysis


In [19]:
# ==============================================================================
# LAB05 — CELL 03: RUN TESTS AND AUDIT THE PROCESSED DATASET
# Objective: Execute the automated tests and audit the configured dataset
# before training the text-classification pipelines.
# ==============================================================================

# Set the NLP sentiment-analysis project as the current working directory.
%cd /content/INF-8239-Ciencia-de-Datos-II/Unit_02_Natural_Language_Processing/nlp_sentiment_analysis

# Run all automated tests stored in the existing test directory.
!python -m pytest test -q

# Run the existing dataset audit script.
!python -m scripts.audit_data

/content/INF-8239-Ciencia-de-Datos-II/Unit_02_Natural_Language_Processing/nlp_sentiment_analysis

==================================== ERRORS ====================================
_______________ ERROR collecting test/test_embeddings_network.py _______________
ImportError while importing test module '/content/INF-8239-Ciencia-de-Datos-II/Unit_02_Natural_Language_Processing/nlp_sentiment_analysis/test/test_embeddings_network.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/usr/lib/python3.13/importlib/__init__.py:88: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
test/test_embeddings_network.py:3: in <module>
    from scripts.embeddings_network import tokenize
scripts/embeddings_network.py:11: in <module>
    from inf8239_u02.config import ROOT, settings
E   ModuleNotFoundError: No module named 'inf8239_u02'
___________________ ERROR collecting test/test_t

In [2]:
%%writefile /content/INF-8239-Ciencia-de-Datos-II/CHANGELOG.md
# Changelog

All notable changes to this repository will be documented in this file.
## [1.6.0] - 2026-10-24

### Added

## [1.6.0] - 2026-10-24

### Added
- Added the complete LAB04 workflow for selecting, downloading, reconstructing, auditing, and validating a public Spanish sentiment corpus.
- Added dataset candidate evaluation based on relevance, license, representativeness, quality, reproducibility, and risk.
- Added SHA-256 verification for the downloaded corpus.
- Added reconstruction of encoded reviews using the provided word dictionary and metadata.
- Added a processed three-class dataset containing positive, neutral, and negative sentiment labels.
- Added automated data-contract tests for required columns and non-empty text values.
- Added dataset documentation, audit results, and Colab portability configuration.
- Updated the repository-level `.gitignore` to exclude local `.env` files and raw and processed datasets from nested projects.

## [1.5.0] - 2026-09-24
### Added
- Created the folder structures for the NLP and Computer Vision projects.
- Created the folder structure for `Unit_02_Natural_Language_Processing/`:
- `app/`
- `data/raw/`
- `data/processed/`
- `docs/`
- `notebooks/`
- `reports/figures/`
- `scripts/`
- `src/`
- `tests/`

## [1.4.0] - 2026-09-24
### Added
- Added additional Green AI visualization outputs under `reports/figures/`.
- Added `model_comparison_f1.png` to compare F1 Macro performance across all evaluated models.
- Added `training_time_comparison.png` to compare median training times.
- Added `inference_time_comparison.png` to compare prediction latency across models.
- Added `model_size_comparison.png` to compare serialized model sizes.
- Added `green_ai_dashboard.png` consolidating F1, training time, inference time, and model size into a single report figure.
- Preserved existing required figures (`pareto.png` and `tsne_two_seeds.png`) and organized all report visualizations in `reports/figures/`.
- Added Green AI benchmarking workflow for U01.LAB03.
- Added six model configurations:
  - Logistic Regression
  - SVM (C=1)
  - SVM (C=10)
  - Random Forest (100 trees)
  - Random Forest (300 trees)
  - Histogram Gradient Boosting
- Added PCA dimensionality-reduction analysis.
- Added t-SNE visualizations using two different random seeds.
- Added Pareto frontier analysis for performance-cost trade-off evaluation.
- Added serialized model size and inference-time benchmarking.
- Added automated Pareto validation tests (`tests/test_green.py`).

### Generated
- Generated `reports/green_ai_results.csv`.
- Generated `reports/figures/pareto.png`.
- Generated `reports/figures/tsne_two_seeds.png`.
- Generated serialized model artifacts under `reports/models/`.

### Documentation
- Added Green AI interpretation and model-selection discussion.
- Added execution environment registration for reproducibility.

## [1.3.0] - 2026-09-22
- Added U01.LAB03 notebook

## [1.2.0] - 2026-09-22

### Added
- Completed the U01.LAB02 dataset audit and baseline modeling workflow.
- Defined the `Class` target and removed potential leakage variables.
- Added preprocessing for numerical and categorical features.
- Compared the `DummyClassifier` baseline against the RBF SVM.
- Added automated dataset-contract tests.
- Added dataset documentation, evaluation results, and conclusions.

### Results
- Dummy baseline F1-Macro: `0.3475`
- SVM F1-Macro: `0.6921`
- SVM accuracy: `0.6926`
- Positive-class recall: `0.6098`
- Dataset contract tests: `3 passed`

### Status
- U01.LAB02 guided workflow completed.
- The justified model adaptation and final Exercise 01 comparison remain pending.

## [1.1.0] - 2026-09-22

### Added
- Started U01.LAB02 public dataset search, selection, and audit workflow.
- Created the dataset profile and documented the analytical problem.
- Compared the Diabetic Retinopathy Debrecen and Dry Bean datasets.
- Applied the dataset acceptance criteria to both candidates.
- Documented dataset sources, targets, classes, dimensions, missing values, and potential leakage risks.
- Added reproducible access to the UCI Machine Learning Repository through the `ucimlrepo` API.
- Added the reusable `download_uci_dataset()` function in `data/data.py`.
- Downloaded and standardized the Diabetic Retinopathy Debrecen dataset as `data/raw/dataset.csv`.
- Added `ucimlrepo` to the project dependencies.

### Changed
- Extended the Exercise 01 notebook to preserve LAB01 as the baseline and begin the LAB02 workflow.
- Updated the data-access process to avoid dependence on personal Google Drive paths.

### Documentation
- Added `docs/dataset_profile.md`.
- Documented the preliminary dataset-selection rationale.
- Added an academic-use disclaimer and a preliminary target-leakage assessment.

### Status
- Reproducible dataset access completed.
- Dataset schema audit, target definition, preprocessing, baseline comparison, and automated tests remain in progress.

## [1.0.0] - 2026-09-21

### Added
- Completed U01.LAB01 guided SVM laboratory using the Breast Cancer dataset.
- Implemented a leakage-safe SVM pipeline with StandardScaler and SVC.
- Added a DummyClassifier baseline for model comparison.
- Performed hyperparameter optimization using GridSearchCV and StratifiedKFold.
- Evaluated model performance using F1-Macro, ROC-AUC, classification reports, and confusion matrices.
- Added reusable build_svm() model factory in src/inf8239_u01/models.py.
- Added unit tests for model validation and pipeline integrity in tests/test_models.py.
- Exported cross-validation results and trained model artifacts.

### Deliverables
- notebooks/01_svm_guiada.ipynb
- src/inf8239_u01/models.py
- tests/test_models.py
- reports/svm_cv_results.csv
- reports/svm_best.joblib

## [0.1.0] - 2026-09-18

### Added
- Established the Unit 01 project structure (LAB00-LAB03 and Exercise_01-Exercise_02).
- Configured baseline environment dependencies in requirements.txt.
- Added .gitignore rules for virtual environments, cache directories, build artifacts, and raw datasets.
- Implemented environment verification modules and corresponding unit tests.
- Created the notebook 00_verificacion.ipynb for environment and interpreter validation.
- Established repository structure for reports, figures, and evidence generation.

### Validated
- Verified automated test execution using pytest.
- Confirmed reproducibility through a local virtual environment setup.
- Validated repository synchronization with the remote GitHub repository.

Overwriting /content/INF-8239-Ciencia-de-Datos-II/CHANGELOG.md


In [ ]:
# ------------------------------------------------------------------------------
# Synchronize the U02.LAB04 notebook with the repository
# ------------------------------------------------------------------------------

from google.colab import _message
from pathlib import Path
import json

# Destination path in the cloned repository
target = Path(
    "/content/INF-8239-Ciencia-de-Datos-II/"
    "Unit_02_Natural_Language_Processing/"
    "notebooks/"
    "U02.LAB04.ipynb"
)

# Verify that the destination folder exists
if not target.parent.is_dir():
    raise FileNotFoundError(
        f"The destination folder does not exist: {target.parent}"
    )

# Retrieve the currently open Colab notebook
nb = _message.blocking_request("get_ipynb")

# Save the notebook in the repository
with target.open("w", encoding="utf-8") as file:
    json.dump(
        nb["ipynb"],
        file,
        ensure_ascii=False,
        indent=1
    )

print("Notebook saved successfully to:")
print(target)

In [10]:
# ==============================================================================
# AUTOMATIC COMMIT AND AUTHENTICATED PUSH
# Objective: Preserve empty directories, stage repository changes, create a
# commit when needed, push the current branch securely, and verify Git status.
# ==============================================================================

# Import os for environment and file-system operations.
import os

# Import stat to configure the temporary authentication script permissions.
import stat

# Import subprocess to execute Git commands and verify their results.
import subprocess

# Import Path for operating-system-independent path management.
from pathlib import Path

# Import userdata to retrieve the GitHub token from Colab Secrets.
from google.colab import userdata


# ------------------------------------------------------------------------------
# 1. DEFINE THE REPOSITORY CONFIGURATION
# ------------------------------------------------------------------------------

# Define the repository root directory.
REPO_PATH = Path(
    "/content/INF-8239-Ciencia-de-Datos-II"
)

# Define the Unit 02 directory.
UNIT_02_PATH = (
    REPO_PATH
    / "Unit_02_Natural_Language_Processing"
)

# Define the GitHub repository information.
GITHUB_USER = "pascual-francisco"
GITHUB_REPO = "INF-8239-Ciencia-de-Datos-II"

# Define the commit message.
COMMIT_MESSAGE = "chore: sync repository updates"

# Retrieve the GitHub token from Colab Secrets.
GITHUB_TOKEN = userdata.get("TOKEN")

# Define the clean GitHub remote URL.
CLEAN_REMOTE_URL = (
    f"https://github.com/{GITHUB_USER}/{GITHUB_REPO}.git"
)


# ------------------------------------------------------------------------------
# 2. VALIDATE THE REPOSITORY AND TOKEN
# ------------------------------------------------------------------------------

# Confirm that the repository directory exists.
if not REPO_PATH.exists():
    raise FileNotFoundError(
        f"The repository was not found:\n{REPO_PATH}"
    )

# Confirm that the repository contains a .git directory.
if not (REPO_PATH / ".git").exists():
    raise FileNotFoundError(
        f"The directory is not a Git repository:\n{REPO_PATH}"
    )

# Confirm that the GitHub token is available.
if not GITHUB_TOKEN:
    raise ValueError(
        "The GitHub token was not found. "
        "Create a Colab Secret named TOKEN and enable notebook access."
    )


# ------------------------------------------------------------------------------
# 3. DEFINE A REUSABLE GIT COMMAND FUNCTION
# ------------------------------------------------------------------------------

def run_git(
    arguments: list[str],
    *,
    check: bool = True,
    capture_output: bool = False,
    environment: dict[str, str] | None = None
) -> subprocess.CompletedProcess:
    """
    Execute a Git command from the repository root.

    Parameters
    ----------
    arguments:
        Git arguments without the initial 'git'.
    check:
        Raise an error when the command returns a nonzero exit code.
    capture_output:
        Capture stdout and stderr instead of displaying them directly.
    environment:
        Optional environment variables used by the Git command.
    """

    return subprocess.run(
        ["git", *arguments],
        cwd=REPO_PATH,
        check=check,
        text=True,
        capture_output=capture_output,
        env=environment
    )


# ------------------------------------------------------------------------------
# 4. PREPARE EMPTY DIRECTORIES
# ------------------------------------------------------------------------------

print("=" * 70)
print("PREPARING EMPTY DIRECTORIES")
print("=" * 70)

if UNIT_02_PATH.exists():

    created_gitkeep_files = []

    # Walk through every directory inside Unit 02.
    for directory_path, directory_names, file_names in os.walk(
        UNIT_02_PATH
    ):
        # A directory is empty when it has no folders and no files.
        if not directory_names and not file_names:
            gitkeep_path = Path(directory_path) / ".gitkeep"
            gitkeep_path.touch()
            created_gitkeep_files.append(gitkeep_path)

    if created_gitkeep_files:
        print(
            f"Created {len(created_gitkeep_files)} "
            ".gitkeep file(s)."
        )
    else:
        print("No empty directories required a .gitkeep file.")

else:
    raise FileNotFoundError(
        f"The Unit 02 directory was not found:\n{UNIT_02_PATH}"
    )


# ------------------------------------------------------------------------------
# 5. DISPLAY THE CURRENT CHANGES
# ------------------------------------------------------------------------------

print("\n" + "=" * 70)
print("CHANGES DETECTED")
print("=" * 70)

# Display modified, deleted, renamed, and untracked files.
run_git([
    "status",
    "--short",
    "--untracked-files=all"
])


# ------------------------------------------------------------------------------
# 6. STAGE THE REPOSITORY CHANGES
# ------------------------------------------------------------------------------

# Stage all changes that are not excluded by .gitignore.
run_git([
    "add",
    "-A"
])

print("\n" + "=" * 70)
print("FILES STAGED FOR COMMIT")
print("=" * 70)

# Retrieve the exact list of staged changes.
staged_result = run_git(
    [
        "diff",
        "--cached",
        "--name-status"
    ],
    capture_output=True
)

staged_output = staged_result.stdout.strip()

# Display the staged changes.
if staged_output:
    print(staged_output)
else:
    print("No files are currently staged.")


# ------------------------------------------------------------------------------
# 7. CREATE A COMMIT ONLY WHEN CHANGES EXIST
# ------------------------------------------------------------------------------

if staged_output:
    print("\nCreating commit...")

    run_git([
        "commit",
        "-m",
        COMMIT_MESSAGE
    ])

    print(f"Commit created: {COMMIT_MESSAGE}")

else:
    print("\nNo new changes to commit.")


# ------------------------------------------------------------------------------
# 8. IDENTIFY THE CURRENT BRANCH
# ------------------------------------------------------------------------------

branch_result = run_git(
    [
        "branch",
        "--show-current"
    ],
    capture_output=True
)

current_branch = branch_result.stdout.strip()

if not current_branch:
    raise RuntimeError(
        "Git could not determine the current branch."
    )


# ------------------------------------------------------------------------------
# 9. CONFIGURE A CLEAN REMOTE URL
# ------------------------------------------------------------------------------

# Keep the permanent origin URL free from credentials.
run_git([
    "remote",
    "set-url",
    "origin",
    CLEAN_REMOTE_URL
])


# ------------------------------------------------------------------------------
# 10. CREATE TEMPORARY NON-INTERACTIVE AUTHENTICATION
# ------------------------------------------------------------------------------

# Define a temporary Git authentication helper.
askpass_path = Path("/tmp/github_colab_askpass.sh")

# Create a temporary script that responds to Git credential prompts.
askpass_content = """#!/bin/sh
case "$1" in
    *Username*) printf '%s\\n' "$GITHUB_USERNAME" ;;
    *Password*) printf '%s\\n' "$GITHUB_TOKEN" ;;
    *) printf '\\n' ;;
esac
"""

# Save the authentication helper.
askpass_path.write_text(
    askpass_content,
    encoding="utf-8"
)

# Allow only the current user to execute or modify the helper.
askpass_path.chmod(
    stat.S_IRUSR
    | stat.S_IWUSR
    | stat.S_IXUSR
)

# Create an environment for the authenticated Git push.
push_environment = os.environ.copy()

# Tell Git which temporary authentication helper to use.
push_environment["GIT_ASKPASS"] = str(askpass_path)

# Prevent Git from opening an interactive credential prompt.
push_environment["GIT_TERMINAL_PROMPT"] = "0"

# Supply the GitHub username only through the process environment.
push_environment["GITHUB_USERNAME"] = GITHUB_USER

# Supply the GitHub token only through the process environment.
push_environment["GITHUB_TOKEN"] = GITHUB_TOKEN


# ------------------------------------------------------------------------------
# 11. PUSH THE CURRENT BRANCH
# ------------------------------------------------------------------------------

print("\n" + "=" * 70)
print("PUSHING TO GITHUB")
print("=" * 70)
print(f"Current branch: {current_branch}")

try:
    # Push the current branch and establish its upstream relationship.
    run_git(
        [
            "push",
            "--set-upstream",
            "origin",
            current_branch
        ],
        environment=push_environment
    )

finally:
    # Delete the temporary authentication helper after the push attempt.
    if askpass_path.exists():
        askpass_path.unlink()


# ------------------------------------------------------------------------------
# 12. DISPLAY THE LATEST COMMITS
# ------------------------------------------------------------------------------

print("\n" + "=" * 70)
print("LATEST COMMITS")
print("=" * 70)

# Display the five most recent commits.
run_git([
    "log",
    "--oneline",
    "--decorate",
    "-5"
])


# ------------------------------------------------------------------------------
# 13. VERIFY THE FINAL SYNCHRONIZATION STATUS
# ------------------------------------------------------------------------------

print("\n" + "=" * 70)
print("FINAL SYNCHRONIZATION STATUS")
print("=" * 70)

# Display the final branch and working-tree status.
run_git([
    "status",
    "--short",
    "--branch"
])

print("\nRepository synchronized successfully.")

PREPARING EMPTY DIRECTORIES
No empty directories required a .gitkeep file.

CHANGES DETECTED

FILES STAGED FOR COMMIT
M	Unit_02_Natural_Language_Processing/notebooks/U02.LAB04.ipynb

Creating commit...
Commit created: chore: sync repository updates

PUSHING TO GITHUB
Current branch: main

LATEST COMMITS

FINAL SYNCHRONIZATION STATUS

Repository synchronized successfully.
